Updated and improved

nrbrook opened the pull request:
I updated the script to work with SDK 15.3.0. I don't think this will work with nRF51 as I'm not sure which soft device to use. It works with nRF52 though.
I also added a macro to import a BLE service and a target to run JLink applications

dariuszseweryn commented on runJLinkExe:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+JLinkExe -device nrf52 -if swd -speed 4000 -autoconnect 1
```
This is hardcoded to `nrf52`. Maybe it would be possible to extract somewhere?

dariuszseweryn commented on runJLinkGDBServer:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+JLinkGDBServer -device nrf52 -strict -timeout 0 -nogui -if swd -speed 1000 -endian little
```
This is hardcoded to `nrf52`. Maybe it would be possible to extract somewhere?

dariuszseweryn commented on readme.md:
````diff
@@ -15,85 +15,71 @@ The script makes use of the following tools:
 	```cmake
 	include_directories(
-	        "${NRF5_SDK_PATH}/components/ble/ble_services/ble_bas"
+	        "${NRF5_SDK_PATH}/<library header directory path>"
````
With these changes the above part does not seem to be adding Battery Service

dariuszseweryn commented on readme.md:
```diff
@@ -15,85 +15,71 @@ The script makes use of the following tools:
+_Note_: you can add `CXX` between `C ASM` to add c++ support
 	
-10. Optionally add additional libraries:
+1. Optionally add additional libraries:
```
Numbering seems off

nrbrook commented:
More fixes. I moved out the compiler settings into a toolchain which can be passed to CMake with `-DCMAKE_TOOLCHAIN_PATH=`, where it then gets the compiler location from the current PATH. Also fixes compiler checking issues and compiling with libraries, and the issues you identified

dariuszseweryn commented:
Hmmm... As for the nRF51 it seems that it is not supported since SDK 12.3.0. That's a bit of a bummer as I took only this chip with me and I will not be able to test on nRF52 against the SDK 15.3.0 till the next week. Sorry for that

dariuszseweryn commented on readme.md:
````diff
@@ -13,87 +13,86 @@ The script makes use of the following tools:
-	```cmake
-	set(NRF_TARGET "nrf51") 
-	```
+	include("cmake-nRF5x/CMake_nRF5x.cmake)
````
```suggestion
	include("cmake-nRF5x/CMake_nRF5x.cmake")
```

dariuszseweryn commented on arm-gcc-toolchain.cmake:
```diff
@@ -0,0 +1,52 @@
+    set(CMAKE_EXE_LINKER_FLAGS_INIT "--specs=nosys.specs")
+endif()
+
+set(TOOLCHAIN_PATH_AND_PREFIX ${ARM_NONE_EABI_TOOLCHAIN_PATH}/${TOOLCHAIN_PREFIX})
```
```suggestion
set(TOOLCHAIN_PATH_AND_PREFIX ${ARM_NONE_EABI_TOOLCHAIN_PATH}/bin/${TOOLCHAIN_PREFIX})
```

dariuszseweryn commented on CMake_nRF5x.cmake:
```diff
@@ -59,116 +54,210 @@ macro(nRF5x_setup)
+            "${NRF5_SDK_PATH}/components/ble/ble_link_ctx_manager"
+            "${NRF5_SDK_PATH}/components/ble/ble_racp"
+            "${NRF5_SDK_PATH}/components/ble/nrf_ble_qwr"
+            "${NRF5_SDK_PATH}/components/ble/ble/peer_manager"
```
```suggestion
            "${NRF5_SDK_PATH}/components/ble/peer_manager"
```

nrbrook replied:
This path was just a mistake when I was updating paths for 15.3.0.

nrbrook commented:
> Would it be possible to update included source files to match the SDK?

I'm not sure what you mean?

> Could you list the things that should be worked on?

Other than nRF51 support and adding more component include macros, I don't think there is anything!

dariuszseweryn commented:
I have updated some macros that were clearly off on my test setup (`examples/ble_peripheral/experimental/ble_app_queued_writes`). Unfortunately I did not have time to go through all the included source files (include macros).

I am wondering how should the `ARM_NONE_EABI_TOOLCHAIN_PATH` should be defined:
1. `/Users/dariuszseweryn/Projects/nRF5x-toolchain/gcc-arm-none-eabi-8-2018-q4-major`
2. `/Users/dariuszseweryn/Projects/nRF5x-toolchain/gcc-arm-none-eabi-8-2018-q4-major/bin`

I would vote for the first option unless I am missing something.

Btw. why despite `arm-gcc-toolchain.cmake` the `CMake_nRF5x.cmake` around line 305 does not use `CMAKE_OBJCOPY` or `CMAKE_SIZE_UTIL`?
```
            COMMAND ${ARM_NONE_EABI_TOOLCHAIN_PATH}/arm-none-eabi-size ${EXECUTABLE_NAME}.out
            COMMAND ${ARM_NONE_EABI_TOOLCHAIN_PATH}/arm-none-eabi-objcopy -O binary ${EXECUTABLE_NAME}.out "${EXECUTABLE_NAME}.bin"
            COMMAND ${ARM_NONE_EABI_TOOLCHAIN_PATH}/arm-none-eabi-objcopy -O ihex ${EXECUTABLE_NAME}.out "${EXECUTABLE_NAME}.hex"
```

nrbrook commented:
> I have updated some macros that were clearly off on my test setup (`examples/ble_peripheral/experimental/ble_app_queued_writes`). Unfortunately I did not have time to go through all the included source files (include macros).

Thanks, I'm not using those myself yet.

> I am wondering how should the `ARM_NONE_EABI_TOOLCHAIN_PATH` should be defined:
> 
>     1. `/Users/dariuszseweryn/Projects/nRF5x-toolchain/gcc-arm-none-eabi-8-2018-q4-major`
> 
>     2. `/Users/dariuszseweryn/Projects/nRF5x-toolchain/gcc-arm-none-eabi-8-2018-q4-major/bin`
> 
> 
> I would vote for the first option unless I am missing something.

As no path other than /bin is used, I changed it to point to that. I don't really have a problem with the first one but it means /bin has to be added to all paths using the variable, and in `arm-gcc-toolchain.cmake` the toolchain search must be modified to point the variable to the parent directory as currently it just finds the directory containing the binary.

> Btw. why despite `arm-gcc-toolchain.cmake` the `CMake_nRF5x.cmake` around line 305 does not use `CMAKE_OBJCOPY` or `CMAKE_SIZE_UTIL`?

I just forgot to update them

dariuszseweryn commented:
> > I am wondering how should the ARM_NONE_EABI_TOOLCHAIN_PATH should be defined:
> >
> > 1. `/Users/dariuszseweryn/Projects/nRF5x-toolchain/gcc-arm-none-eabi-8-2018-q4-major`
> > 
> > 2. `/Users/dariuszseweryn/Projects/nRF5x-toolchain/gcc-arm-none-eabi-8-2018-q4-major/bin`
> > I would vote for the first option unless I am missing something.
>
> As no path other than /bin is used, I changed it to point to that. I don't really have a problem with the first one but it means /bin has to be added to all paths using the variable, and in arm-gcc-toolchain.cmake the toolchain search must be modified to point the variable to the parent directory as currently it just finds the directory containing the binary.

The first option would make the script more compatible with the previous version of the script

Excuse me if I ask too many/noob questions — I am not too familiar with compiling C stuff nor CMake. Is there a reason for separating `nRF5x_toolchainSetup` in a macro where it needed by `nRF5x_setup`?

nrbrook commented:
> The first option would make the script more compatible with the previous version of the script
I have renamed the variable and added compatibility for the old variable
> Excuse me if I ask too many/noob questions — I am not too familiar with compiling C stuff nor CMake. Is there a reason for separating `nRF5x_toolchainSetup` in a macro where it needed by `nRF5x_setup`?
The toolchain must be set up before calling `project`. `nRF5x_setup` creates targets which I think implicitly creates a project if one does not already exist. If the toolchain is set up after a project call, CMake detects the change in compiler and reruns configuration, so it gets stuck in a loop.

dariuszseweryn commented:
> The toolchain must be set up before calling project. nRF5x_setup creates targets which I think implicitly creates a project if one does not already exist. If the toolchain is set up after a project call, CMake detects the change in compiler and reruns configuration, so it gets stuck in a loop.

I see. 

I have also tried to run `cmake --build "cmake-build" --target START_JLINK` against my setup with nRF52 Dev Kit `PCA10056` but it seems that either `JLinkGDBServer` or `JLinkExe` cannot connect. Not sure though if this is related to the script somehow

nrbrook commented:
Try running the scripts individually to see which doesn't work. I'm using PCA10040 but there is nothing specific to this in the scripts.

dariuszseweryn commented:
It seems that `JLinkGDBServer` and `JLinkExe` cannot be run concurrently

nrbrook commented:
It works for me on a Mac...try changing the speed

dariuszseweryn commented:
Tried both at speed 1000 (kHz). I run on MacBook Pro 15" — mid 2014. Feel free to suggest other tests
For me it is not a deal breaker as I did not yet used any of those programs directly in this setup

I plan to play with this PR for a bit and merge to `master` afterwards. Previous commit will start a branch for `SDK12`, and the `master` will be `SDK15` branch as well so other potential users could find integration points easily. Maybe I will backport some changes as well as I need to make some research against both nRF51 and nRF52.

dariuszseweryn commented:
Hello @nrbrook 

I have merged your PR. Thank you for all your work it will be very helpful for some :)

Cheers!

nrbrook commented:
Thanks, and thanks for making the project! Nordic development is much nicer in CLion. I'm not sure why the utilities are working for me but not you. I guess you will have updated them to the latest versions. Could be a difference in the boards perhaps, I'm using PCA10040.

dariuszseweryn commented:
@nrbrook Someone recently pointed out that the default `ALL` target tries to execute all `FLASH_ERASE`, `FLASH_SOFTDEVICE` and `FLASH_{your target name}` in parallel where they should be run serially. I have tried to make it happen this way by making some wrappers and adding `DEPENDS` clause but it did not work for me.

I am thinking whether it is possible to make them run serially in case of the default target being build or should I rather remove all custom targets from `ALL` target? WDYT?

nrbrook commented:
I think they should probably not be part of `ALL` including the JLink targets, but the build targets should be.
